### ISSUE TITLE
Handle Location and Plaque events correctly

### ISSUE THREAD

**Issue by nemerle** [Segs/Segs#136]
### Expected behavior
Replicate the original game logic here : this needs more information 
### Actual behavior/symptoms
Server only prints debug info whenever a Location or Plague is visited

**Comment by broxen** [comment]
This process should look like:
- Click upon Plaque
- Client opens Modal box showing Plaque Text and "OK" button
- Pull plaque text from LUA script, so that these can be customized per-server. These should probably be in a larger badges file, unless for non-badge plaques, like the one under Atlas, which should probably pull from a specific LUA file? Perhaps one that also includes other zone settings, like map name, atmosphere, day/night/cycle etc. This scope could be expanded later as necessary.
- If plaque is part of a History Badge, set flag that it has been read.
- If all plaques in history badge have been completed, award badge.

**Notes:**
- I can't find a single screenshot or video showing this behavior anywhere. I've asked on reddit to see if anyone has one laying around.
- Badges and History plaques were added starting in I2 of CoH, more info here [ParagonWiki - Badges by Issue](https://paragonwiki.com/wiki/Badges_by_Issue)
- The text for these plaques may have been found in Badges.bin later, but I can't confirm with any of my copies of the client
- I know the badge locations are stored in the I2+ map files, so it may be easiest to pull directly from there, though future implementations should allow for customization in scripts, so you can set your own exploration badges, etc. This would be harder to do with plaques, as they have an object in the game world, but perhaps could still be done?

**Comment by nemerle** [comment]
Great work broxen, thanks.
As for badges/location visited LUA side events, I was actually thinking of implementing those, basically on map server startup, the server would load a "specific_map_name/locations.lua" file, where we have location_name
```lua
locations_dict = {}
locations_dict[0x12345678] = "Atlas statue"

function location_visited(location_name_hash,x,y,z)
  client.admin_chat_message("You have visited"..locations_dict[location_name_hash])
end
```
As for badges, I think we're going to have to postpone it until we move to I2 :)

**Comment by broxen** [comment]
That sounds great!

I may have misunderstood. Are you referring to the `Entering Atlas Plaza` and `Entering Prometheus Park` text when entering different neighborhoods? Because I forgot all about that but you can see it in action in this video: https://www.youtube.com/watch?v=juH9pgJ4Ik4

**Comment by nemerle** [comment]
I don't know actually, just that the client will send LocationVisited messages as well as PlaqueVisited, it is likely that LocationVisited is used by "Entering Location Name".

**Comment by broxen** [comment]
If LocationVisited correlates to the neighborhood boundaries on the maps then definitely.

Those neighborhoods show up on in-game maps, but you can see them here as well: https://paragonwiki.com/w/images//e/e1/Atlas_Park_VidiotMap.png

**Comment by nemerle** [comment]
One of the reasons I'm not sure it's because the client's internal logic looks like this
```cpp
static QSet<QString> s_visitedLocations;

void reportVisited(GroupDef *node, Vector3 *pos)
{
    QString result;
    QString name;
    if ( !visitLocationIgnored(node) )
    {
        name = node->getPropertyValue("VisitLocation");
        if ( !name.isEmpty() && !s_visitedLocations.contains(name))
        {
            command_stream.SendPackedBits(1, MapEventTypes::evLocationVisited);
            command_stream.SendString(g_input_pak, name);
            command_stream.SendF32(pos->x);
            command_stream.SendF32(pos->y);
            command_stream.SendF32(pos->z);
        }
    }
    name = node->getPropertyValue("Plaque");
    if ( !name.isEmpty() )
    {
        command_stream.SendPackedBits( 1, MapEventTypes::evPlaqueVisited);
        command_stream.SendString( name);
        command_stream.SendF32( pos->x);
        command_stream.SendF32( pos->y);
        command_stream.SendF32( pos->z);
    }
    name = node->getPropertyValue("Kiosk");
    if ( !name.isEmpty() )
    {
        sendConsoleCommand(QString("kiosk %1 home").arg(name));
    }
}
```

**Comment by nemerle** [comment]
But maybe VisitLocation property is set on nodes that group many of the smaller level ones, thus encompassing big areas of the map..

**Comment by broxen** [comment]
That's possible, or maybe it's early Exploration Badge logic?

Where can I add qdebug to output when this is sent by the client? I'll run around to known badge locations and neighborhood perimeters to see if there's any discernible effect.

The Kiosk appears to be the CityInfo Terminal near Atlas Statue, but these weren't functional until I2 as well: https://paragonwiki.com/wiki/Issue_2

**Comment by nemerle** [comment]
The most recent commit should already contain the following code in MapInstance.
```cpp
void MapInstance::on_location_visited(LocationVisited *ev)
{
    qWarning() << "Unhandled location visited event:" << ev->m_name <<
                  QString("(%1,%2,%3)").arg(ev->m_pos.x).arg(ev->m_pos.y).arg(ev->m_pos.z);
}

void MapInstance::on_plaque_visited(PlaqueVisited * ev)
{
    qWarning() << "Unhandled plaque visited event:" << ev->m_name <<
                  QString("(%1,%2,%3)").arg(ev->m_pos.x).arg(ev->m_pos.y).arg(ev->m_pos.z);
}
```

**Comment by broxen** [comment]
I'll test tonight and report any findings.

**Comment by nemerle** [comment]
Great, thank You.

**Comment by broxen** [comment]
**Test results:**
Entering a new map neighborhood results in the text `Entering [area]` automatically. This appears to be handled by the client.

**Plaque under Atlas + Police Boxes**
Clicking on the plaque under Atlas outputs this:
```
Warning : Unhandled location visited event: "Statue1" "(126.5,27.4315,-300)"
Warning : Unhandled plaque visited event: "CityCenterFemaleStatue" "(126.5,27.4315,-300)"
```
which matches up with the `/loc` command in the client, but the place names make no sense to me.

But also, clicking on the yellow police boxes around town produced output:
```
Warning : Unhandled location visited event: "Patrol_Easy_5_City_01_01" "(1099.5,0,124.5)"
Warning : Unhandled plaque visited event: "PatrolBox" "(1099.5,0,124.5)"
```
So clearly these are the same type (as it's the only type implemented) and my guess is that the `location` references the file where the text is stored, and the `plaque` references what "type" the object is.

**_It might be beneficial to include the type number in the debug output_**

Some other tests:

**Known Badge Locations**
Testing known badge locations (top of city hall, top of atlas statue) to see if they would trigger on_location_visited appears to do nothing.

**Known Plaque Locations**
Testing known plaque locations from I2 results in no effect.

**Info Kiosk**
Clicking on the community information terminal located at `219.2  23.6  -237.1` near Atlas statue, and the other one near the Monorail results in:
```
Console command received kiosk home home
```
and the chat output `[Local]StrongDude: home home`

**Door to City Hall, Sewers, or other Doors**
Clicking on the door to City Hall resulted in:
```
Unhandled command event type 9
Debug   : Not all bits were consumed
```

**Saving Config**
Saving client config resulted in:
```
Unhandled command event type 65
Debug   : Not all bits were consumed
```

**Clicking on Character**
Clicking on another character results in:
```
Unhandled command event type 32
```

**Character Info**
Selecting `Actions>Info` for another character results in:
```
Unhandled command event type 56
Debug   : Not all bits were consumed
```

**Selecting Location on Map**
Selecting one of the locations marked on the map (like the Monorail station) results in:
```
Unhandled command event type 11
Debug   : Not all bits were consumed
```

**Selecting ID from the menu**
Selecting ID from the user menu results in:
```
Unhandled command event type 67
Debug   : Not all bits were consumed
```

The server crash in the output is because I pressed `ctrl+c` in the server window.

**Files:**
[output.log](https://github.com/Segs/Segs/files/1452535/output.log)
[crash.log](https://github.com/Segs/Segs/files/1452538/crash.log)

**Comment by broxen** [comment]
Also pressing `Z` results in:
```
Unhandled command event type 33
Debug   : Not all bits were consumed
```
The `z` key is bound to `/powexec_unqueue and Cancels the queued power. Bound to the Z key by default. ` according to the interwebs. The options menu doesn't show any default binding for that key.

**Comment by broxen** [comment]
Expanding the Inspiration Window using the up arrow results in:
```
Unhandled command event type 17
Debug   : Not all bits were consumed
```

**Comment by nemerle** [comment]
Great scott, that's a lot of packet types :D
I'll start chipping away at those today/tomorrow

**Comment by broxen** [comment]
Let me know how/where I can help.

**Comment by nemerle** [comment]
Well, those should be all of those except for packet 32, which I think has something to do with aborting 'current' power vs 33 being 'abort queued' power.

**Comment by nemerle** [comment]
As for Location/Plaque issue, it might be that Location is a 'template' that should be loaded, and Plaque contains the actual values to be filled in the template, then the resulting rich-text string is built and sent to the client ?

**Comment by broxen** [comment]
This is an amazing start to tackling all of this!

I think you're on the right track for plaques, however based upon the police boxes, where plaque had the same type consistently, but location name changed based upon `/loc` it seems like it's reverse of what you're thinking:

That is to say that plaque appears to be the actual template and location the specific values. This can be tested by running around to various police boxes and clicking on them, which results in a different location name each time.

For packet 32, I only rec'd that when clicking on another player, so I can't say for sure why it would be some kind of abort power method. My first guess was actually some kind of update for the selected entities stats. This may save overhead by eliminating updates until you can actually "see" the data (such as origin/archetype/health/etc)

It would be updated consistently for a Team member, as you'll see their health bar constantly.

Let me know how I can help test, or if you decide to shelf this to focus on other items.

Side note: is there any way currently to change zones? Slash command `/movemap` or anything?

**Comment by nemerle** [comment]
re: Map change
The current design/thought is like this:
* each game server is connected to a map server router
* map server router manages N map server processes
* each map server process has methods `initializeMap(name)` and `shutDown(redirectTo)`, `transferTargetReady(name,target)`, `getTransferToken(for_player)`
* when Map server starts, it is in 'idle' state, and no players can connect to it
* when Map server receives `initializeMap(name)' command, it will load the required map data, and notify the map server router it is ready by `mapServerReady(id,name,...)`
* when Map server router receives `mapServerReady` it registers the server in the routing map for that name, and notify all other map servers that
* when a request is made to enter a specific map, the router will search in the routing maps, and if all servers in the map are `busy` (high number of players ), it will try to initialize one of the 'idle' maps to a specific map. When the target map is ready, the map server router will send `transferTargetReady`, when source map server  receives `transferTargetReady` it will request a transfer token for each player waiting on the transfer list. when a transfer token is received, player finally gets a mapTransfer packet from it's source map server.

-----
Sadly, currently we have no map server router and 1 map server process with hardcoded map-file :smile: , so there's a lot of work till we have proper map transfers :)

**Comment by broxen** [comment]
This makes perfect sense. Thank you for the explanation and for all of your hard work!

**Comment by nemerle** [comment]
We now have ability to run simple scripts whenever user interacts with a plaque/location, example files are located in Projects/CoX/Data/MapInstances/City_00_01/
```lua
function location_visited(location_hash)
    return "Something something"
end
```
```lua
function plaque_visited(plaque_hash)
    client:simple_dialog("You have visited a plaque!")
    return "I've informed the client"
end
```

**Comment by broxen** [comment]
This is great!

Leaving this here for reference:
https://paragonwiki.com/wiki/Plaque

**Comment by mobbyg** [comment]
Oh yeah!! 
![](https://i.imgur.com/LXkrxyU.gif)

**Comment by mobbyg** [comment]
OK, so now that this is a thing, how do you associate a plaque location with the server? And then the text of said plaque?

**Comment by broxen** [comment]
@mobbyg I think what you're getting at is how do we determine a specific Plaque's Hash for `location_hash` and `plaque_hash`, and I have this question as well.

If your question is more-so about adding new, custom plaques to the game world, then I don't think the current implementation handles that use-case. Right now, when you click on a plaque in the game world, the client sends a series of packets to the server saying essentially "hey, I'm a plaque located at these coordinates, tell me what to do." Segs is currently taking that packet info and providing a system for response, but it's a very basic system.

@nemerle can speak more to this issue, as he's much more knowledgeable than me.

**Comment by mobbyg** [comment]
@broxen Yeah, the existing plaques is what I meant. Like the Atlas one which is the only one on the map I believe till you jump over to Galaxy City which isn’t a thing right now anyways. 

While I can’t help in the coding area when it comes to C++, I can do some in game content to have something to flesh out a basic server. 

And I am looking at learning LUA to help there. If anyone knows of some video tutorials or if Lynda has something, I can jump on that. 

But it could also be time for that Learn C in 21 days book. LOL

**Comment by nemerle** [comment]
Server reports the name of location/plaque alongs the hash values.

I'm starting to lean towards making the locations/plaques data non-LUA

**Comment by mobbyg** [comment]
@Nemerle as in not making it scriptable?

**Comment by nemerle** [comment]
The scripts will still be called, but the data will be stored in text files along the scripts, and text parsing and filling in placeholders in the templates will be done c++ side

**Comment by broxen** [comment]
I would script `plaques.lua` more like below. This is assuming locations work the way I think they do in relation to plaques.

```lua
do
  local location = {
    Statue1 = {126.5,27.4315,-300},
    Statue2 = {100,20,-100}, //whatever the m_name/coords are for this
  }
  local plaque = {
    CityCenterFemaleStatue = "This statue commemorates the heroic death of the mighty Atlas. On December 7th, 1941, the Forces of the original 5th Column launched a surprise attack on Paragon City. Atlas single-handedly fought back a large group of advancing 5th Column soldiers, and prevented them from gaining a foothold beyond Independence Port. His brave sacrifice held the area until the Freedom Phalanx arrived and drove back the invaders.",
    GalaxyGirlStatue = "Kelly Graham never thought she was destined for greatness. That changed one night when she was walking home from work. She came upon an innocent woman being dragged into an alley by some masked men. Kelly called out for help, to no avail. As her fear and anger grew, she launched herself toward the alley. After a blur of flying, energy blasting, and healing light... Galaxy Girl was born!",
  }

  function plaque_visited(plaque, location)
    for _,v in ipairs(plaque) do
      if player:location() = location then
        local dlg_text = plaque[v] or "This plaque is not implemented yet!"
        // open dialog with plaque text
        client:simple_dialog(dlg_text)
      end
    end
  end

end

// Call our function using the hash values provided by the server
plaque_visited{plaque_hash, location_hash}
```

**Comment by broxen** [comment]
Some testing results (using latest git, with no changes)

```
Debug   : Attempting a call to script location_visited with: "Statue1" 2901444586
Warning : Unhandled location visited event: "Statue1" "(126.5,27.4315,-300)"
Debug   : Attempting a call to script plaque_visited with: "CityCenterFemaleStatue" 3805458902
Warning : Unhandled plaque visited event: "CityCenterFemaleStatue" "(126.5,27.4315,-300)"
Received window state 28 - 4
Received window state 28 - 4
Debug   : Attempting a call to script location_visited with: "Statue1" 2901444586
Warning : Unhandled location visited event: "Statue1" "(126.5,27.4315,-300)"
Debug   : Attempting a call to script plaque_visited with: "CityCenterFemaleStatue" 3805458902
Warning : Unhandled plaque visited event: "CityCenterFemaleStatue" "(126.5,27.4315,-300)"
Received window state 28 - 4
Received window state 28 - 4
Received window state 28 - 4
Unhandled command event type 32
Received window state 21 - 1
Warning : Unhandled entity info requested 1
Received window state 21 - 2
Received window state 21 - 2
Received window state 21 - 2
Received window state 21 - 3
Received window state 21 - 4
Received window state 24 - 4
Received window state 24 - 4
Received window state 24 - 4
Received window state 21 - 1
Warning : Unhandled entity info requested 1
Received window state 21 - 2
Received window state 21 - 3
Received window state 21 - 4
Console command received friend TestGuy
Console command received trade TestGuy
Console command received t TestGuy, test
Received window state 6 - 1
Console command received i TestGuy
Received window state 6 - 2
Received window state 6 - 3
Received window state 6 - 4
Received window state 24 - 4
Received window state 24 - 4
Unhandled command event type 32
pressed
Received window state 21 - 1
Warning : Unhandled entity info requested 2
```

Couple of things:
- `Unhandled command event type 32` still showing when clicking upon another character
- Entity Archetype not updating correctly for some reason my Defender was showing as a Blaster. I'll test further and open an issue if necessary.

**Comment by broxen** [comment]
Moving this to v0.7.0 as it seems like there might be some interest from Discord channel.

In particular, this can provide a rough framework to build off of when implementing further Lua scripts
